docs: :memo: revise package guide to "local-first" (#1109)
seedcase-project/seedcase-sprout: docs/guide/packages.qmd

## Creating a data package

We've designed Sprout to be used in a similar way that Git repositories
or Python virtual environments are used (for instance, as we recommend
in the [installation guide](installation.qmd)). What that means is, we
assume and expect that you will be creating and managing a data package
in the root of your Python (or Git) project. The same folder where your
`.git/` folder is or your `pyproject.toml` file is will also be the same

> folder that has your `datapackage.json` file. With this expectation and

Author: Hmm, I don't think that belongs here. That might be in the design section, since it is more of an "explanation" type doc than a "how-to".

design, many of Sprout's helper path functions assume that the working
directory is the same directory where the `datapackage.json` file is (or
will be) stored and where the `pyproject.toml` file (or `.git/` folder)
is.

With that in mind, let's make our first data package! A data package
always needs a `datapackage.json` file. This file contains a set of

> properties, or metadata, about the data package and, eventually, about

Author: Yea, I want to make that glossary repo and than go through all our repos and link to the glossary. But that's later since we don't have that set up yet.

the data resources within it. To set up this `datapackage.json` file, we
start with the `create_package_properties()` function. This function
takes the properties you want to add to the data package and the path
where you want to store the data package as arguments. So first, we need

> to establish our properties.

Reviewer: OK there seems to be a switching back and forth between “you” and “we” further down as well. Maybe this should be aligned :)

> to establish our properties.

Author: This is a guide, so it is "you", not "we". If it's a tutorial, it could be "we". Totally agree though, it needs to be aligned in this doc.

The "we" words could be replaced with "Sprout has", I will do that after committing your suggestions.

    properties = properties,
    path = package_path
)
```

::: callout-important
The `create_package_properties()` function will give an error if the
`PackageProperties` object is missing some of its required fields or if
they are not filled in correctly. In that case, a `datapackage.json`
file won't be created.
:::

This creates the initial structure of your new package. The
`create_package_properties()` function created the `datapackage.json`
file in your data package `diabetes-study` folder, which contains the
properties you added to it. The newly created file would be:

```{python}
#| echo: false
print(package_path.glob("**/*"))
```

<!-- TODO: Add section on building the README -->

## Editing package properties

If you made a mistake and want to update the properties in the current
`datapackage.json`, you can use the `edit_package_properties()` function
while using the helper `path_properties()` function to point to the
`datapackage.json` file. The default behavior of the `path_properties()`
function is to look in the current working directory, but for this
guide, it is pointing to the temporary folder created from above.

```{python}
sp.edit_package_properties(

>   path=sp.path_properties(path=package_path),

Author: I realized this doesn't build now because it depends on #1114

> pprint(properties)

Reviewer: I just noticed that the ID is empty at this point. Is it added when using `create_package_properties()`? At least there’s an ID in the "Editing package properties" section. Maybe a comment on when/how that is added would be helpful.

```

Now, let's create our data package with these properties.

::: callout-note
For this guide, we will create this data package in a temporary folder.
In a real project, you would create the data package in the root of your

> project. We'll make this temporary folder using:

Author: Once there are more example repos that @K-Beicher completes, the idea is to include a callout block on the Guide landing page to link to that "gallery of examples" for them to look at how it's done in a real project.

```{python}
import tempfile
from pathlib import Path

temp_path = tempfile.TemporaryDirectory()
package_path = Path(temp_path.name) / "diabetes-study"

# Create the path to the package
package_path.mkdir(parents=True)
```
:::

```{python}
package_path = sp.create_package_properties(
    properties = properties,

>     path = package_path

Author: Not used in local settings.

object and will not create the `datapackage.json` file.
:::

This creates the initial structure of your new package. The
`create_package_properties()` function created the `datapackage.json`
file in your data package `diabetes-study` folder, which contains the

> properties you added to it.

Author: Nice idea!

> properties you added to it.

Author: Unfortunately, Python doesn't seem to have a built-in tree function. So I will just list the files in the folder.

You now have the basic starting point for adding data resources to your
data package.

> ## Making a package in a multi-user environment

Reviewer: ```suggestion
## Creating a data package in a multi-user server environment
```
To align with previous header (“*Creating* a *data* package”) and the first sentence below (L160).

Reviewer: Should the header “Creating a data package” have something added to it like “ … in a local environment” (I don’t necessarily like that specific phrasing)?

Author: I think it might be nice to have a separate guide for the multi-user environment to keep things contained. For now, we can leave it here so we can see how it looks like. I made an issue about this.

If you are making and managing data packages in a multi-user server

> environment that will have or has multiple data packages, there are some

Reviewer: Does that mean that if someone know they will have multiple data packages (locally), they should also follow this part of the guide?

Author: Not really. This is really only for multi-user settings.

> very small changes you can make to creating and managing packages. While
> Sprout assumes the functions are being used to create and manage a data
> package in the working directory, all the functions take the `path` as
> an argument. Which means, they can run in any directory. With that in

Reviewer: ```suggestion
very small changes you can make to creating and managing packages. Since 
all the functions to create and manage data packages in Sprout take the `path` as an argument, they can run in any directory. With that in
```

mind, we have a series of helper path functions that can be used to
point to a "global" Sprout storage location.

The first thing you'll need to decide is where you want to store your
data packages in this type of environment. By default, Sprout will
create it in `~/sprout/packages/` on Linux (see
[Outputs](/docs/design/interface/outputs.qmd) for operating system
specific locations), but you can change this by setting the
`SPROUT_GLOBAL` environment variable. For instance, maybe you want the
location to be `~/Desktop/sprout/` or `~/Documents`. For our example, we

> will store it in our current working directory in the hidden folder

Reviewer: ```suggestion
will store our data package in the current working directory in a hidden folder called
```

Reviewer: Why in a hidden folder, actually?

Author: Doesn't have to be hidden. I can fix that to be not hidden. It just tends to be common in Linux/Unix (which includes Mac) that global folders and files for programs are stored in the user's home as hidden folders.

`.storage/`.

```{python}
import os
os.environ["SPROUT_GLOBAL"] = ".storage/"
```

> Now, we can create a new data package in this global location, with only

Reviewer: ```suggestion
Now, we can create a new data package using this global variable, with only
```
